### PR TITLE
expression: fine-grained precision infer when casting const to Decimal

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/30961

Problem Summary:

### What is changed and how it works?

When the constant is converted to Decimal type, precision and scale can be completely determined.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fine-grained precision infer when casting const to Decimal
```

### CODE REVIEW

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -1937,7 +1937,24 @@ func WrapWithCastAsDecimal(ctx sessionctx.Context, expr Expression) Expression {
 	}
 	types.SetBinChsClnFlag(tp)
 	tp.Flag |= expr.GetType().Flag & mysql.UnsignedFlag
-	return BuildCastFunction(ctx, expr, tp)
+	castExpr := BuildCastFunction(ctx, expr, tp)
```
Comment (windtalker) [inline expression/builtin_cast.go:1940]: Will `WrapWithCastAsDecimal` be used when user explicit cast a value to decimal using `cast(xx as decimal)`?

Comment (wshwsh12) [reply expression/builtin_cast.go:1940]: No. WrapWithCastAsDecimal don't need the target field type, and only used by implicit cast.
Explicit cast will use BuildCastFunction directly.
And cast(xx as decimal) is equivalent to cast(xx as decimal(10,0))...

**Review by windtalker — APPROVED** [review]
LGTM

**Review by guo-shaoge — APPROVED** [review]

**Pull request merged**